README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
hrIPW
=====

This function allows estimating the log hazard ratio associated with a
binary exposure using a Cox PH model weighted by the propensity score.
Propensity model is estimated using a simple logistic regression.
Variance estimation takes into account the propensity score estimation
step with the method proposed by Hajage *et al.* (2018)
\doi{10.1002/bimj.201700330}. Both the average treatment effect on the
overall (ATE) or the treated (ATT) population can be estimated. For ATE,
both unstabilized and stabilized weights can be used. Ties are handled
throught the Breslow approximation.

Example
-------

``` r
library(hrIPW)

## Using a simulated cohort
data(hrData, package = "hrIPW")
hrIPW(hrData, time = "time", status = "status", exposure = "Trt",
      variables = paste("X", 1:9, sep = ""), wtype = "ATE-stab")
#> $coefficient
#> [1] -0.01717167
#> 
#> $std
#> [1] 0.03282054
#> 
#> $ciinf
#> [1] -0.08149875
#> 
#> $cisup
#> [1] 0.04715541
#> 
#> $p.value
#> [1] 0.6008359

# Standard error could be compared with the (robust) Lin's standard error
# which does not take into account the propensity score estimation step:
modT <- glm(Trt ~ X1 + X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9, data = hrData, family = "binomial")
probT <- predict(modT, type = "response")
hrData$w <- 1/ifelse(hrData$Trt == 1, probT, 1 - probT)
library(survival)
coxph(Surv(time, status) ~ Trt + cluster(id), data = hrData, method = "breslow", weights = w)
#> Call:
#> coxph(formula = Surv(time, status) ~ Trt + cluster(id), data = hrData, 
#>     weights = w, method = "breslow")
#> 
#>        coef exp(coef) se(coef) robust se     z    p
#> Trt -0.0174    0.9827   0.0142    0.0354 -0.49 0.62
#> 
#> Likelihood ratio test=1.52  on 1 df, p=0.2
#> n= 10000, number of events= 10000

# or with the bootstrap-based standard-error (see Austin 2016):
f.boot <- function(data, i, wtype) {
    df <- data[i, ]
    modT <- glm(Trt ~ X1 + X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9, data = df, family = "binomial")
    probT <- predict(modT, type = "response")
    df$w <- 1/ifelse(df$Trt == 1, probT, 1 - probT)
    
    return(coxph(Surv(time, status) ~ Trt, data = df, weights = w)$coef)
}

library(boot); set.seed(1234)
#> 
#> Attaching package: 'boot'
#> The following object is masked from 'package:survival':
#> 
#>     aml
rcoefs <- boot(data = hrData, statistic = f.boot, R = 500)$t
sd(rcoefs)
#> [1] 0.03283661

## Using the DIVAT data base (package IPWsurvival, to be installed)
data(DIVAT, package = "IPWsurvival")
hrIPW(data = DIVAT, time = "times", status = "failures", exposure = "ecd",
      variables = c("age", "hla", "retransplant"), wtype = "ATE-unstab")
#> $coefficient
#> [1] 0.4661664
#> 
#> $std
#> [1] 0.2564084
#> 
#> $ciinf
#> [1] -0.03638477
#> 
#> $cisup
#> [1] 0.9687177
#> 
#> $p.value
#> [1] 0.0690546

# Standard error could be compared with the (robust) Lin's standard error
# which does not take into account the propensity score estimation step:
modT <- glm(ecd ~ age + hla + retransplant, data = DIVAT, family = "binomial")
probT <- predict(modT, type = "response")
DIVAT$w <- 1/ifelse(DIVAT$ecd == 1, probT, 1 - probT)
DIVAT$id <- 1:nrow(DIVAT)
coxph(Surv(times, failures) ~ ecd + cluster(id), data = DIVAT, method = "breslow", weights = w)
#> Call:
#> coxph(formula = Surv(times, failures) ~ ecd + cluster(id), data = DIVAT, 
#>     weights = w, method = "breslow")
#> 
#>       coef exp(coef) se(coef) robust se   z     p
#> ecd 0.4650    1.5920   0.0517    0.2579 1.8 0.071
#> 
#> Likelihood ratio test=83.28  on 1 df, p=<2e-16
#> n= 1837, number of events= 582

# or with the bootstrap-based estimation (see Austin 2016):
f.boot2 <- function(data, i, wtype) {
    df <- data[i, ]
    modT <- glm(ecd ~ age + hla + retransplant, data = df, family = "binomial")
    probT <- predict(modT, type = "response")
    df$w <- 1/ifelse(df$ecd == 1, probT, 1 - probT)
    
    return(coxph(Surv(times, failures) ~ ecd, data = df, weights = w)$coef)
}

set.seed(1234)
rcoefs2 <- boot(data = DIVAT, statistic = f.boot2, R = 500)$t
sd(rcoefs2)
#> [1] 0.2885173
```


NEWS.md
# hrIPW 0.1.3

## Issues

* fix no more existing suggested package


R/binipw.r
# binIPW <- function(data, outcome, exposure, variables, wtype, estimate, alpha = 0.05) {
#
#     n <- nrow(data)
#     X <- as.matrix(cbind(1, data[, variables]))
#     fct <- function(x) t(rbind(x)) %*% rbind(x)
#
#     formT <- as.formula(paste(exposure, "~", paste(variables, collapse = " + ")))
#     modT <- glm(formT, data = data, family = "binomial")
#     data$pT <- predict(modT, type = "response")
#     if (wtype[1] == "ATE") {
#         data$w <- 1/ifelse(data[, exposure] == 1, data$pT, 1 - data$pT)
#     } else if (wtype[1] == "ATT") {
#         data$w <- ifelse(data[, exposure] == 1, 1, data$pT/(1 - data$pT))
#     }
#
#     A1 <- sum(data$w*data[, exposure]*data[, outcome])
#     B1 <- sum(data$w*data[, exposure])
#     P1 <- A1/B1
#     A0 <- sum(data$w*(1-data[, exposure])*data[, outcome])
#     B0 <- sum(data$w*(1-data[, exposure]))
#     P0 <- A0/B0
#
#     x <- t(apply(X, 1, fct))
#     pTx <- sweep(x, 1, data$pT*(1-data$pT), "*")
#     mpTx <- array(colSums(pTx), dim=c(length(vars)+1, length(vars)+1))
#     mpTxinv <- solve(mpTx)
#
#     if (wtype[1] == "ATE") {
#         y1 <- sweep(X, 1, data[, exposure]*(1/data$pT-1), "*")
#         y1 <- sweep(y1, 1, data[, outcome] - P1, "*")
#         my1 <- colSums(y1)
#         gamma11 <- mpTxinv %*% my1
#         xgamma11 <- X%*%gamma11
#         xgamma11 <- sweep(xgamma11, 1, data[, exposure]-data$pT, "*")
#         E1 <- ((data[, outcome] - P1)*data[, exposure]/data$pT - xgamma11)
#
#         y0 <- sweep(X, 1, (1-data[, exposure])*(data$pT/(1-data$pT)), "*")
#         y0 <- sweep(y0, 1, data[, outcome] - P0, "*")
#         my0 <- colSums(y0)
#         gamma10 <- mpTxinv %*% my0
#         xgamma10 <- X%*%gamma10
#         xgamma10 <- sweep(xgamma10, 1, data[, exposure]-data$pT, "*")
#         E0 <- ((data[, outcome] - P0)*(1-data[, exposure])/(1-data$pT) + xgamma10)
#     } else if (wtype[1] == "ATT") {
#         E1 <- data[, exposure]*(data[, outcome]-P1)/(mean(data[, exposure]))
#
#         nt <- sum((1-data[, exposure])*data$pT/(1-data$pT))/n
#         y0 <- sweep(X, 1, (1-data[, exposure])*(data$pT/(1-data$pT)), "*")
#         y0 <- sweep(y0, 1, data[, outcome] - P0, "*")
#         my0 <- colSums(y0)
#         gamma20 <- mpTxinv %*% my0
#         xgamma20 <- X%*%gamma20
#         xgamma20 <- sweep(xgamma20, 1, data[, exposure]-data$pT, "*")
#         E0 <- ((data[, outcome] - P0)*(1-data[, exposure])*data$pT/(1-data$pT) + xgamma20)/(nt)
#     }
#
#     if (estimate[1] == "log(OR)") {
#         coefficient <- logit(P1) - logit(P0)
#         E0.or <- E0/(P0*(1-P0))
#         E1.or <- E1/(P1*(1-P1))
#         U.or <- E1.or - E0.or
#         std <- sqrt(var(U.or)/n)
#     } else if (estimate[1] == "log(RR)") {
#         coefficient <- log(P1) - log(P0)
#         E0.rr <- E0/(P0)
#         E1.rr <- E1/(P1)
#         U.rr <- E1.rr - E0.rr
#         std <- sqrt(var(U.rr)/n)
#     } else if (estimate[1] == "RD") {
#         coefficient <- P1 - P0
#         U.rd <- E1 - E0
#         std <- sqrt(var(U.rd)/n)
#     }
#
#     ci <- coefficient + qnorm(c(alpha/2, 1-alpha/2))*c(std)
#     p.value <- 2*(1-pnorm(abs(coefficient/std)))
#
#     return(list(coefficient = coefficient, std = std, ciinf = ci[1], cisup = ci[2], p.value = p.value))
# }
#
# contIPW <- function(data, outcome, exposure, variables, wtype, alpha = 0.05) {
#     binIPW(data, outcome, exposure, variables, wtype, estimate = "RD", alpha = alpha)
# }


R/hripw.r
#' Hazard ratio estimation using Cox model weighted by the propensity score
#'
#' This function allows estimating the log hazard ratio associated with a binary exposure using a Cox PH model weighted by the propensity score. Propensity model is estimated using a simple logistic regression. Variance estimation takes into account the propensity score estimation step with the method proposed by Hajage *et al.* (2018) \doi{10.1002/bimj.201700330}. Both the average treatment effect on the overall (ATE) or the treated (ATT) population can be estimated. For ATE, both unstabilized and stabilized weights can be used. Ties are handled through the Breslow approximation.
#'
#' @param data The data.frame to be analyzed.
#' @param time A character string indicating the name of the follow up times column.
#' @param status A character string indicating the name of the failure column (0=right censored, 1=event).
#' @param exposure A character string indicating the name of the exposure column (0=unexposed, 1=exposed).
#' @param variables A character vector indicating the names of the confounders to be accounted for in the propensity model.
#'   Only numeric variables could be handled, so factors, characters and other eligible classes should be expanded into dummy/indicator variables.
#' @param wtype A character string indicating the type of weights that should be used
#'   ('ATE-unstab' for unstabilized ATE weights, 'ATE-stab' for stabilized ATE weights, or 'ATT' for ATT weights)
#' @param alpha A numeric value indicating the (bilateral) alpha level used for the computation of the confidence interval
#'
#' @return A list with the following elements:
#' * coefficient: the log-HR associated with the exposure
#' * std: the standard error
#' * ciinf and cisup: lower and upper limits of the (1-alpha) condidence interval
#' * p.value: the p-value
#' @md
#'
#' @importFrom stats as.formula complete.cases glm pnorm predict qnorm var
#' @importFrom survival coxph Surv
#' @export
#' @references
#’   Hajage D, Chauvet G, Belin L, Lafourcade A, Tubach F, De Rycke Y
#'   Closed-form variance estimator for weighted propensity score estimators with survival outcome.
#'   Submitted to Statistics in Medicine (2017).
#'
#'   Austin PC. Variance estimation when using inverse probability of treatment weighting (IPTW) with survival analysis.
#'   Statistics in medicine, 30;35(30):5642-5655, 2016. <doi: 10.1002/sim.7084>
#'
#'   Lin DY, Wei LJ. The Robust Inference for the Cox Proportional Hazards Model.
#'   Journal of the American Statistical Association 84(408):1074-1078, 1989. <doi: 10.1080/01621459.1989.10478874>
#' @examples
#' ## Using a simulated cohort
#' data(hrData, package = "hrIPW")
#' hrIPW(hrData, time = "time", status = "status", exposure = "Trt",
#'       variables = paste("X", 1:9, sep = ""), wtype = "ATE-stab")
#'
#' # Standard error could be compared with the (robust) Lin's standard error
#' # which does not take into account the propensity score estimation step:
#' modT <- glm(Trt ~ X1 + X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9, data = hrData, family = "binomial")
#' probT <- predict(modT, type = "response")
#' hrData$w <- 1/ifelse(hrData$Trt == 1, probT, 1 - probT)
#' library(survival)
#' coxph(Surv(time, status) ~ Trt + cluster(id), data = hrData, method = "breslow", weights = w)
#'
#' # or with the bootstrap-based standard-error (see Austin 2016):
#' \dontrun{
#' f.boot <- function(data, i, wtype) {
#'     df <- data[i, ]
#'     modT <- glm(Trt ~ X1 + X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9, data = df, family = "binomial")
#'     probT <- predict(modT, type = "response")
#'     df$w <- 1/ifelse(df$Trt == 1, probT, 1 - probT)
#'
#'     return(coxph(Surv(time, status) ~ Trt, data = df, weights = w)$coef)
#' }
#'
#' library(boot); set.seed(1234)
#' rcoefs <- boot(data = hrData, statistic = f.boot, R = 500)$t
#' sd(rcoefs)
#' }
#'
#' ## Using the DIVAT data base (package IPWsurvival, to be installed)
#' data(dataDIVAT2, package = "RISCA")
#' hrIPW(data = dataDIVAT2, time = "times", status = "failures", exposure = "ecd",
#'       variables = c("age", "hla", "retransplant"), wtype = "ATE-unstab")
#'
#' # Standard error could be compared with the (robust) Lin's standard error
#' # which does not take into account the propensity score estimation step:
#' modT <- glm(ecd ~ age + hla + retransplant, data = dataDIVAT2, family = "binomial")
#' probT <- predict(modT, type = "response")
#' dataDIVAT2$w <- 1/ifelse(dataDIVAT2$ecd == 1, probT, 1 - probT)
#' dataDIVAT2$id <- 1:nrow(dataDIVAT2)
#' coxph(Surv(times, failures) ~ ecd + cluster(id), data = dataDIVAT2, method = "breslow", weights = w)
#'
#' # or with the bootstrap-based estimation (see Austin 2016):
#' \dontrun{
#' f.boot <- function(data, i, wtype) {
#'     df <- data[i, ]
#'     modT <- glm(ecd ~ age + hla + retransplant, data = df, family = "binomial")
#'     probT <- predict(modT, type = "response")
#'     df$w <- 1/ifelse(df$ecd == 1, probT, 1 - probT)
#'
#'     return(coxph(Surv(times, failures) ~ ecd, data = df, weights = w)$coef)
#' }
#'
#' set.seed(123)
#' rcoefs <- boot(data = dataDIVAT2, statistic = f.boot, R = 200)$t
#' sd(rcoefs)
#' }
hrIPW <- function(data, time, status, exposure, variables, wtype, alpha = 0.05) {

    data <- data[, c(time, status, exposure, variables)]
    na <- !complete.cases(data)
    if (any(na)) {
        warning(paste(sum(na), "observation(s) dropped."))
        data <- data[!na, ]
    }

    formT <- as.formula(paste(exposure, "~", paste(variables, collapse = " + ")))
    modT <- glm(formT, data = data, family = "binomial")
    data$pT <- predict(modT, type = "response")
    if (wtype[1] == "ATE-unstab") {
        data$w <- 1/ifelse(data[, exposure] == 1, data$pT, 1 - data$pT)
    } else if (wtype[1] == "ATE-stab") {
        data$w <- ifelse(data[, exposure] == 1, mean(data[, exposure]), 1 - mean(data[, exposure]))/ifelse(data[, exposure] == 1, data$pT, 1 - data$pT)
    } else if (wtype[1] == "ATT") {
        data$w <- ifelse(data[, exposure] == 1, 1, data$pT/(1 - data$pT))
    }

    formSurv <- as.formula(paste("Surv(", time, ", ", status, ") ~", exposure))
    coefficient <- unname(coxph(formSurv, data = data, weights = data$w)$coefficients)

    fct <- function(x) t(rbind(x)) %*% rbind(x)
    data <- data[order(data[, time]),]
    n <- nrow(data)
    X <- as.matrix(cbind(1, data[, variables]))

    wegT <- data$w*exp(coefficient*data[, exposure])
    TwegT <- data[, exposure]*wegT
    s0 <- (sum(wegT) - cumsum(wegT) + wegT)/n
    s1 <- (sum(TwegT) - cumsum(TwegT) + TwegT)/n
    R <- s1/s0

    R2 <- sapply(seq_along(data[, time]), function(i) mean((data$w*data[, status]*as.numeric(data[, time][i] >= data[, time])/s0)*(data[, exposure][i] - R)))

    XX <- t(apply(X, 1, fct))
    pTXX <- sweep(XX, 1, data$pT*(1 - data$pT), '*')
    A <- array(colSums(pTXX), dim = c(length(variables) + 1, length(variables) + 1))/n
    Ainv <- solve(A)

    U0 <- data[, status]*(data[, exposure] - R) - exp(coefficient*data[, exposure])*R2

    if (wtype[1] == "ATE-unstab") {
        c <- 0
        bint <- (-data[, exposure]*(1 - data$pT)/data$pT + (1 - data[, exposure])*data$pT/(1 - data$pT))*U0
    } else if (wtype[1] == "ATE-stab") {
        c <- mean((data[, exposure]/data$pT - (1 - data[, exposure])/(1 - data$pT))*U0)
        prev <- mean(data[, exposure])
        bint <- (-data[, exposure]*prev*(1 - data$pT)/data$pT + (1 - data[, exposure])*data$pT*(1 - prev)/(1 - data$pT))*U0
    } else if (wtype[1] == "ATT") {
        c <- 0
        bint <- ((1 - data[, exposure])*data$pT/(1 - data$pT))*U0
    }
    bint <- sweep(X, 1, bint, '*')
    bint <- colSums(bint)/n
    b <- Ainv %*% bint
    B <- mean(data$w*data[, status]*R*(1 - R))
    U <- (data$w*U0 + c*(data[, exposure] - data$pT) + sweep(X, 1, data[, exposure] - data$pT, '*')%*%b)/B
    U <- U[, 1]

    std <- sqrt(var(U)/n)
    ci <- coefficient + qnorm(c(alpha/2, 1-alpha/2))*c(std)
    p.value <- 2*(1-pnorm(abs(coefficient/std)))

    return(list(coefficient = coefficient, std = std, ciinf = ci[1], cisup = ci[2], p.value = p.value))
}

